src/tools/r2rdump/R2RDump.cs

```diff
@@ -0,0 +1,200 @@
+                }
+
+                Size = curOffset - Offset;
+            }
+
+            public void Dump()
+            {
+                Console.WriteLine("============== R2R Header ==============");
+                Console.WriteLine("Signature: 0x{0:X8} ({1})", Signature, SignatureString);
+                Console.WriteLine("RelativeVirtualAddress: 0x{0:X8}", RelativeVirtualAddress);
+                if (Signature == READYTORUN_SIGNATURE) {
+                    Console.WriteLine("Size: " + Size);
+                    Console.WriteLine("MajorVersion: 0x{0:X4}", MajorVersion);
+                    Console.WriteLine("MinorVersion: 0x{0:X4}", MinorVersion);
+                    Console.WriteLine("Flags: 0x{0:X8}", Flags);
+                    Console.WriteLine("NumberOfSections: " + NumberOfSections);
+                    for (int i = 0; i < NumberOfSections; i++)
+                    {
+                        Console.WriteLine("------------------");
+                        Sections[i].Dump();
+                    }
+                }
+            }
+        }
+
+        enum OutputTypes
+        {
+            Error = 0x00000000,
+            IsR2R = 0x00000001,
+            Machine = 0x00000002,
+            Filename = 0x00000003,
+        }
+
+        static void Output<T>(T data, OutputTypes type)
+        {
+            switch (type)
+            {
+                case OutputTypes.Error:
+                    Console.WriteLine("Error: " + data);
+                    break;
+                case OutputTypes.IsR2R:
+                    Console.WriteLine("IsR2R: " + data);
+                    break;
+                case OutputTypes.Machine:
+                    Console.WriteLine("Machine: " + data);
+                    break;
+                case OutputTypes.Filename:
+                    Console.WriteLine("FileName: " + data);
+                    break;
+            }
+        }
+
+        static long GetField(ref int start, int size)
+        {
+            byte[] bytes = new byte[size];
+            Array.Copy(pe, start, bytes, 0, size);
+            start += size;
+
+            if (size == 8)
+            {
+                return BitConverter.ToInt64(bytes, 0);
+            }
+            else if (size == 4)
+            {
+                return BitConverter.ToInt32(bytes, 0);
+            }
+            else if (size == 2)
+            {
+                return BitConverter.ToInt16(bytes, 0);
+            }
+            return bytes[0];
+        }
+
+        static void Main(string[] args)
+        {
+            try
+            {
+                if (args.Length < 1)
+                {
+                    throw new System.ArgumentException("File name must be passed as argument");
+                }
+                string filename = args[0];
+                Output<String>(filename, OutputTypes.Filename);
+
+                pe = File.ReadAllBytes(filename);
+                Stream fs = File.OpenRead(filename);
```

Comment:
Also, you should either use only one of `File.ReadAllBytes` or `File.OpenRead` so that the file is read just once. The current code is opening and reading the file twice.